### PR TITLE
storage: prevent removed replica from getting lease

### PR DESCRIPTION
**Pull request by tbg**
Out for early review since the fix seems easier than what has been proposed
in #15385 so far. Will add tests in this PR.

Only review my commits, those of @andreimatei are from #15711 (which is
waiting for @bdarnell as well).

---

Make sure that a lease with an intended lease holder that has since been
removed catches a forcedError.

See #15385.

### CODE REVIEW

**Comment by petermattis** [comment]
:lgtm:, though we need @bdarnell's eyes on this as well.

---

Reviewed 1 of 1 files at r1.
Review status: 0 of 31 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15754#-:-Kj_UN-kLSWqt0BCeXP9:b7x2zv3)*

File: pkg/storage/replica.go [hunk 1]
```diff
@@ -980,6 +980,34 @@ func (r *Replica) redirectOnOrAcquireLease(ctx context.Context) (LeaseStatus, *r
...
+						// all proposals, processRaftCommand checks that their ProposalLease is
+						// compatible with the active lease for the log position. For commands
+						// proposed on the lease holder, the command queue then serializes
+						// everything. But lease commands get created on followers based on their
+						// local state and thus without being sequenced through the command queue.
+						// Thus a recently removed follower (unaware of its own removal) could
+						// submit a proposal for the lease (correctly using as a ProposerLease the
+						// last active lease), and would receive it given the up-to-date
+						// ProposerLease. Hence, an extra check is in order: processRaftCommand
+						// makes sure that leases for a replica not in the descriptor are bounced.
```
Comment (a-robinson) [inline pkg/storage/replica.go]: I think this would be a little more clear if you replaced s/leases/lease requests/

**Review by a-robinson — APPROVED** [review]
LGTM, and thanks for independently re-thinking through all this.

**Comment by bdarnell** [comment]
:lgtm:

This change downstream of raft could theoretically cause replica divergence during rolling upgrades, but since it's an assertion against an event that would already cause the range to get stuck (and does not appear to be occurring in practice), it should be OK. 

---

Reviewed 3 of 4 files at r6, 1 of 1 files at r7.
Review status: 4 of 31 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15754#-:-KjcdKZKO1TeGuJ_uIrY:bny263c)*

**Comment by andreimatei** [comment]
:lgtm:

---

Review status: 4 of 31 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15754#-:-Kjdterr_5RXcOg92cvz:bnfp4nl)*

File: pkg/storage/replica.go (hunk 1 above)
Comment (tschottdorf) [inline pkg/storage/replica.go]: Done.

**Comment by tschottdorf** [comment]
Updated with a test. Spent nontrivial time on a higher-level test, but I think this one's as straightforward as we're gonna get.

**Comment by bdarnell** [comment]
LGTM

---

Reviewed 3 of 27 files at r8, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15754#-:-KkWdzhOjnNiTd1-SfLZ:bb6ywmx)*

**Pull request merged**